test/consul_helpers.go

```diff
@@ -0,0 +1,196 @@
+		defer gcp.DeleteImage(t, gcpProjectID, imageID)
+	})
+
+	test_structure.RunTestStage(t, "deploy", func() {
+		// GCP only supports lowercase names for some resources
+		uniqueID := strings.ToLower(random.UniqueId())
+		serverClusterName := fmt.Sprintf("consul-server-cluster-%s", uniqueID)
+		clientClusterName := fmt.Sprintf("consul-client-cluster-%s", uniqueID)
+		gcpProjectID := gcp.GetGoogleProjectIDFromEnvVar()
+		gcpZone := test_structure.LoadString(t, exampleFolder, SavedGCPZone)
+		imageID := test_structure.LoadArtifactID(t, exampleFolder)
+
+		terraformOptions := &terraform.Options{
+			TerraformDir: exampleFolder,
+			Vars: map[string]interface{}{
+				ConsulClusterExampleVarProject:              gcpProjectID,
+				ConsulClusterExampleVarZone:                 gcpZone,
+				ConsulClusterExampleVarServerClusterName:    serverClusterName,
+				ConsulClusterExampleVarClientClusterName:    clientClusterName,
+				ConsulClusterExampleVarServerClusterTagName: serverClusterName,
+				ConsulClusterExampleVarClientClusterTagName: clientClusterName,
+				ConsulClusterExampleVarServerSourceImage:    imageID,
+				ConsulClusterExampleVarClientSourceImage:    imageID,
+				ConsulClusterExampleVarServerClusterSize:    ConsulClusterExampleDefaultNumServers,
+				ConsulClusterExampleVarClientClusterSize:    ConsulClusterExampleDefaultNumClients,
+			},
+			//EnvVars: map[string]string{
+			//	GoogleProjectIDEnvVar: gcpProjectID,
+			//},
+		}
+		test_structure.SaveTerraformOptions(t, exampleFolder, terraformOptions)
+
+		terraform.InitAndApply(t, terraformOptions)
+	})
+
+	test_structure.RunTestStage(t, "validate", func() {
+		gcpProjectID := gcp.GetGoogleProjectIDFromEnvVar()
+		gcpZone := test_structure.LoadString(t, exampleFolder, SavedGCPZone)
+		terraformOptions := test_structure.LoadTerraformOptions(t, exampleFolder)
+
+		// Check the Consul servers
+		checkConsulClusterIsWorking(t, ConsulClusterExampleOutputServerInstanceGroupName, terraformOptions, gcpProjectID, gcpZone)
+
+		// Check the Consul clients
+		checkConsulClusterIsWorking(t, ConsulClusterExampleOutputClientInstanceGroupName, terraformOptions, gcpProjectID, gcpZone)
+	})
+}
+
+// Check that the Consul cluster comes up within a reasonable time period and can respond to requests
+func checkConsulClusterIsWorking(t *testing.T, groupNameOutputVar string, terratestOptions *terraform.Options, projectID string, zone string) {
+	groupName := terraform.OutputRequired(t, terratestOptions, groupNameOutputVar)
+
+	// It can take a few minutes for the managed instance group to boot up
+	maxRetries := 30
+	timeBetweenRetries := 5 * time.Second
+
+	// Check every 5 seconds until an instance has joined the managed instance group
+	nodeIPAddress := retry.DoWithRetry(t, fmt.Sprintf("Waiting for instances in group %s", groupName), maxRetries, timeBetweenRetries, func() (string, error) {
+		ip, err := getIPAddressOfManagedInstance(t, projectID, zone, groupName)
```

Comment:
The function name `getIPAddressOfManagedInstance` is a little confusing because it implies that a Managed Instance Group has a single IP address, versus a collection of Instances, each of which has an IP address. I'd rather the interface you use to accomplish this look like:

```golang
randNodeIndex := random.Random(1, ConsulClusterExampleDefaultNumServers)
instance := gcp.GetFromInstanceGroup(t, projectID, zone, groupName, randNodeIndex)
ip := gcp.GetPublicIPOfInstance(t, projectID, zone, instance)
```

This will also have the added benefit of testing a random server in the Managed Instance Group each time versus always testing the first.

You can also wrap the above in its own `getRandomPublicIpFromInstanceGroup` function.

Reply:
Sure, that works.

Reply:
I played around a bit with the (somewhat awkward) [GCP API](https://cloud.google.com/compute/docs/reference/rest/v1/), and now I realize that the API likes to pass around Instance IDs, not the whole instance object. 

Therefore, I think the following code makes the most sense, which won't require any upstream Terratest dependencies:

```golang
...
randNodeIndex := random.Random(1, ConsulClusterExampleDefaultNumServers)
instanceIDs := gcp.GetInstanceIdsForInstanceGroup(t, projectID, zone, groupName)
instanceID := instanceIDs[randNodeIndex]
ip := gcp.GetPublicIPOfInstance(t, projectID, zone, instanceID)

return ip, nil
...
```

You might also consider checking that `randNodeIndex` is less than `len(InstanceIDs)`. Sorry that my initial feedback led you astray a bit here.

Reply:
@josh-padnick ok let me make the changes. Yes, the GCP API is pretty confusing in some places.